### PR TITLE
refactor: Improve Promise Handling, Validation, and Code Efficiency Across Scripts

### LINKED ISSUE #1402: Refactor and Bug Fixes for Promise Handling and Validation
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/GDevelopApp/GDevelop-extensions/issues)

### Enter the name of the extension

Promise Handling & Validation Refactor

### Describe the bug of the extension

The current implementation of some key functions in the extension (such as verifyExtension.js and extract-extension.js) lacks consistent error handling and validation. This can lead to unhandled promise rejections, improper error reporting, and potential crashes when incorrect inputs are provided.

### Steps to reproduce

1. Go to the extension's folder.
2. Attempt to load a zip file or JSON file with invalid names or structures.
3. Observe inconsistent error handling or crashes.
4. In some cases, file extraction might fail silently or without proper error messages.

### GDevelop platform

Desktop

### GDevelop version

5.1.160

### Platform info

<details>
OS: Windows, Linux, macOS
OS Version: Windows 10, macOS 11.0
Browser (for Web): Chrome, Firefox
Device (for Mobile): N/A

</details>

### Additional context

---

The refactor applies consistent `async/await` promise handling, improves type checking, and combines redundant loops in various scripts:
- **`verifyExtension.js`**: Introduced a `try-catch` block for proper async error handling, added logging, and improved the flow of promise-based error management.
- **`extract-extension.js`**: Wrapped operations in a `try-catch` block to streamline promise handling, preventing unhandled rejections.
- **`ExtensionNameValidator.js`**: Added string validation for the extension name to avoid runtime errors from invalid inputs.
- **`PascalCase.js`**: Combined loops to reduce code duplication and added type checking to ensure extension names are valid strings.

This improves the overall robustness of the extension and prevents crashes or silent failures in several key processes.

---

### PR DESCRIPTION
**Pull request by Viole07**
---

### **Pull Request Description :**

Closes #1402

This pull request refactors and improves several scripts in the project by enhancing error handling, improving validation logic, and streamlining promise management. Below are the key changes made across different files:

1. **`verifyExtension.js`:**
   - Refactored promise handling by introducing a `try-catch` block for consistent async error handling across the function.
   - Replaced the mix of `.catch()` and `async/await` with a more streamlined approach, ensuring any unhandled async errors are caught and logged properly.
   - Added detailed error logging to help with debugging during extension verification.

2. **`extract-extension.js`:**
   - Refactored to streamline the promise handling by wrapping the entire function in a `try-catch` block.
   - This ensures that any errors related to loading or extracting the zip file are handled in a single place, providing more consistent error handling.
   - Improved the error logging by including the actual error message in the returned result for better context during debugging.

3. **`ExtensionNameValidator.js`:**
   - Added a type check for `extensionName` to ensure it is a string before proceeding with further validation.
   - This prevents potential runtime errors when invalid data types are passed, making the function more robust.

4. **`PascalCase.js`:**
   - Combined two similar loops into a single loop that processes both `publicEventsFunctions` and `eventsBasedBehaviors`. This reduces code duplication and simplifies the logic.
   - Added a type check for `extension.name` to ensure it is a string before validating the PascalCase, preventing errors when the name is not provided or is of the wrong type.

### **Benefits:**
- Improved code readability and maintainability by consolidating promise handling and reducing redundancy.
- Increased robustness through added type checks and better error logging, making debugging easier.
- Simplified logic by combining redundant loops and refactoring validation functions into smaller, reusable components.

---

### CODE REVIEW

**Comment by Viole07** [comment]
@4ian , why is this check not passing.. i tried fixing the issue using 

npx prettier --write "scripts/**/*.js" "scripts/**/*.ts"

and it's still not working..

File: scripts/lib/rules/PascalCase.js [hunk 1]
```diff
@@ -29,9 +29,19 @@ async function validate({
...
+
   if (legacyCamelCaseExtensions.has(name)) return;
-  for (const { name } of publicEventsFunctions) checkPascalCase(name, onError);
-  for (const { name } of eventsBasedBehaviors) checkPascalCase(name, onError);
+
+  // Combine both loops into one to reduce duplication
+  const allFunctions = [...publicEventsFunctions, ...eventsBasedBehaviors];
+  for (const { name } of allFunctions) {
+    checkPascalCase(name, onError);
+  }
```
Comment (arthuro555) [inline scripts/lib/rules/PascalCase.js:44]: This offers no new benefit, does an inefficient copy, and unnecessarily complexifies the logic.

File: scripts/lib/rules/PascalCase.js [hunk 2]
```diff
@@ -29,9 +29,19 @@ async function validate({
   publicEventsFunctions,
   onError,
 }) {
+  // Check if extension name is a valid string
+  if (typeof name !== 'string') {
+    onError('Invalid extension name. Expected a string.');
+    return;
+  }
```
Comment (arthuro555) [inline scripts/lib/rules/PascalCase.js:36]: Typechecking is already statically enforced via typescript: this is unnecessary complexity.

File: scripts/lib/ExtensionNameValidator.js [hunk 3]
```diff
@@ -1,5 +1,8 @@
 /** @param {string} extensionName */
 exports.isValidExtensionName = (extensionName) => {
+  // Ensure extensionName is a string to prevent runtime errors
+  if (typeof extensionName !== 'string') return false;
+
```
Comment (arthuro555) [inline scripts/lib/ExtensionNameValidator.js:5]: Typechecking is already statically enforced via typescript: this is unnecessary complexity.

File: scripts/extract-extension.js [hunk 4]
```diff
@@ -15,40 +15,31 @@ exports.extractExtension = async function (
...
     await pipeline(
       file.nodeStream(),
       createWriteStream(`${extensionsFolder}/community/${file.name}`)
     );
+    return { extensionName };
   } catch (e) {
-    console.warn(`JSZip extraction error caught: `, e);
-    return { error: 'zip-error' };
+    console.warn(`JSZip loading or extraction error caught: `, e);
+    return { error: 'zip-error', details: e.message };
```
Comment (arthuro555) [inline scripts/extract-extension.js:43]: The try catch block should only cover the block it is handling errors for (the `JSZip.loadAsync` call)

File: scripts/check-single-extension.js [hunk 5]
```diff
@@ -18,46 +18,47 @@ exports.verifyExtension = async function (extensionName, options) {
...
+    if (validationDetails.length > 0)
+      return {
+        code: 'rule-break',
+        errors: validationDetails.map(({ message }) => message),
+      };
+    return { code: 'success' };
+  } catch (error) {
+    // General error catch for any unhandled async errors in the process
+    console.error(`Error verifying extension: ${extensionName}`, error);
+    return { code: 'not-found' }; // Return a default error code, you could also return something more specific based on the context
```
Comment (arthuro555) [inline scripts/check-single-extension.js:62]: Most definitely not! The error code determines what error message is sent back to the submitter - sending them a message about a file not being found is not generic, and goes against the point of providing feedback. The script crashing is already a handled case, so this unnecessarily strips the context of the error.

**Review by arthuro555 — CHANGES_REQUESTED** [review]

**Comment by arthuro555** [comment]
Since every proposed change here does not in fact improve promise handling, validation or code efficiency, I'll close this PR.

**Pull request closed without merging**